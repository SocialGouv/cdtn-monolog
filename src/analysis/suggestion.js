import * as DatasetUtil from "../dataset";
import * as DataForge from "data-forge";

const reportType = "suggestion";

const analyse = (dataset) => {
  // get unique suggestion selection events in visits
  const uniqueSuggs = DataForge.DataFrame.concat(
    DatasetUtil.getVisits(dataset)
      .select((v) => DatasetUtil.toUniqueSuggestions(v))
      .where((v) => v.count() > 0)
      .toArray()
  ).deflate((r) => r.suggestionSelection);

  const total = uniqueSuggs.count();

  const normalisation = (c) => 100 * Math.sqrt(c / total);
  const round = (n) => Math.round((n + Number.EPSILON) * 100) / 100;

  // count selections for each suggestion
  const counts = uniqueSuggs
    .groupBy((value) => value)
    .select((suggGroup) => ({
      suggestion: suggGroup.first(),
      count: suggGroup.count(),
      weight: round(normalisation(suggGroup.count())),
      reportType,
    }))
    .inflate()
    .orderByDescending((r) => r.weight);

  return counts.toArray();
};

export { analyse, reportType };
