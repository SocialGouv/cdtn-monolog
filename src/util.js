export const SERVICE_URL = "https://code.travail.gouv.fr/";

export const urlToPath = (url) => {
  if (url.startsWith(SERVICE_URL)) {
    return url.slice(SERVICE_URL.length);
  } else {
    return url;
  }
};

export const actionTypes = {
  visitType: "visit_content",
  selectSuggestionType: "select_suggestion",
};
