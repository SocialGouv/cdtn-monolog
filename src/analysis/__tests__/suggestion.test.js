import * as Suggestion from "../suggestion";
import * as util from "../../util";
import { DataFrame } from "data-forge";

describe("Suggestion", () => {
  const toSuggestionSelectionAction = (uvi, suggestionSelection) => ({
    uvi,
    suggestionSelection,
    type: util.actionTypes.selectSuggestionType,
  });

  const generate = (a) =>
    new DataFrame(a.map(([uvi, ss]) => toSuggestionSelectionAction(uvi, ss)));

  it("should be unique per visit", () => {
    const actions = generate([
      [1, "q1"],
      [1, "q1"],
      [1, "q2"],
    ]);
    expect(Suggestion.analyse(actions)).toMatchSnapshot();
  });

  it("should be weighted properly", () => {
    const actions1 = generate([
      [1, "q1"],
      [1, "q2"],

      [2, "q1"],
      [2, "q2"],

      [3, "q2"],

      [4, "q1"],
    ]);
    expect(Suggestion.analyse(actions1)).toMatchSnapshot();

    const actions2 = generate([
      [1, "q1"],
      [1, "q2"],
      [2, "q1"],
    ]);
    expect(Suggestion.analyse(actions2)).toMatchSnapshot();
  });
});
