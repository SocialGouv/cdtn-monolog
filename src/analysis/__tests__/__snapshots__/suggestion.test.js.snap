// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestion should be unique per visit 1`] = `
Array [
  Object {
    "count": 1,
    "reportType": "suggestion",
    "suggestion": "q1",
    "weight": 70.71,
  },
  Object {
    "count": 1,
    "reportType": "suggestion",
    "suggestion": "q2",
    "weight": 70.71,
  },
]
`;

exports[`Suggestion should be weighted properly 1`] = `
Array [
  Object {
    "count": 3,
    "reportType": "suggestion",
    "suggestion": "q1",
    "weight": 70.71,
  },
  Object {
    "count": 3,
    "reportType": "suggestion",
    "suggestion": "q2",
    "weight": 70.71,
  },
]
`;

exports[`Suggestion should be weighted properly 2`] = `
Array [
  Object {
    "count": 2,
    "reportType": "suggestion",
    "suggestion": "q1",
    "weight": 81.65,
  },
  Object {
    "count": 1,
    "reportType": "suggestion",
    "suggestion": "q2",
    "weight": 57.74,
  },
]
`;
